File: nexuscli-go/nexus_upload.go
package main

import (
	"fmt"
	"io"
	"mime/multipart"
	"net/http"
	"os"
	"path/filepath"
	"strings"

	"github.com/schollz/progressbar/v3"
)

// UploadOptions holds options for upload operations
type UploadOptions struct {
	Logger    Logger
	QuietMode bool
}

func collectFiles(src string) ([]string, error) {
	var files []string
	err := filepath.Walk(src, func(path string, info os.FileInfo, err error) error {
		if err != nil {
			return err
		}
		if !info.IsDir() {
			files = append(files, path)
		}
		return nil
	})
	return files, err
}

func uploadFiles(src, repository, subdir string, config *Config, opts *UploadOptions) error {
	filePaths, err := collectFiles(src)
	if err != nil {
		return err
	}
	// Calculate total bytes to upload
	totalBytes := int64(0)
	fileSizes := make([]int64, len(filePaths))
	for i, filePath := range filePaths {
		info, err := os.Stat(filePath)
		if err != nil {
			return err
		}
		fileSizes[i] = info.Size()
		totalBytes += info.Size()
	}

	// Create progress bar - write to /dev/null when disabled
	showProgress := isatty() && !opts.QuietMode
	progressWriter := os.Stdout
	if !showProgress {
		progressWriter, _ = os.OpenFile(os.DevNull, os.O_WRONLY, 0)
	}
	bar := progressbar.NewOptions64(totalBytes,
		progressbar.OptionSetWriter(progressWriter),
		progressbar.OptionShowBytes(true),
		progressbar.OptionSetDescription("Uploading bytes"),
	)

	pr, pw := io.Pipe()
	writer := multipart.NewWriter(pw)

	// Write multipart form in a goroutine
	errChan := make(chan error, 1)
	go func() {
		defer pw.Close()
		for idx, filePath := range filePaths {
			relPath, _ := filepath.Rel(src, filePath)
			relPath = filepath.ToSlash(relPath)
			f, err := os.Open(filePath)
			if err != nil {
				errChan <- err
				return
			}
			defer f.Close()
			part, err := writer.CreateFormFile(fmt.Sprintf("raw.asset%d", idx+1), filepath.Base(filePath))
			if err != nil {
				errChan <- err
				return
			}
			reader := io.TeeReader(f, bar)
			if _, err := io.Copy(part, reader); err != nil {
				errChan <- err
				return
			}
			_ = writer.WriteField(fmt.Sprintf("raw.asset%d.filename", idx+1), relPath)
		}
		if subdir != "" {
			_ = writer.WriteField("raw.directory", subdir)
		}
		writer.Close()
		errChan <- nil
	}()

	uploadEndpoint := fmt.Sprintf("%s/service/rest/v1/components?repository=%s", config.NexusURL, repository)
	req, err := http.NewRequest("POST", uploadEndpoint, pr)
	if err != nil {
		return err
	}
	req.SetBasicAuth(config.Username, config.Password)
	req.Header.Set("Content-Type", writer.FormDataContentType())
	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()
	if goroutineErr := <-errChan; goroutineErr != nil {
		return goroutineErr
	}
	if resp.StatusCode == 204 {
		opts.Logger.Printf("Uploaded %d files from %s\n", len(filePaths), src)
	} else {
		respBody, _ := io.ReadAll(resp.Body)
		opts.Logger.Printf("Failed to upload files: %d %s\n", resp.StatusCode, string(respBody))
	}
	return nil
}

func uploadMain(src, dest string, config *Config, opts *UploadOptions) {
	repository := dest
	subdir := ""
	if strings.Contains(dest, "/") {
		parts := strings.SplitN(dest, "/", 2)
		repository = parts[0]
		subdir = parts[1]
	}
	err := uploadFiles(src, repository, subdir, config, opts)
	if err != nil {
		fmt.Println("Upload error:", err)
		os.Exit(1)
	}
}

File: nexuscli/nexus_upload.py
import os
import sys
import requests
from typing import Optional, List
from tqdm import tqdm
from requests_toolbelt.multipart.encoder import MultipartEncoder, MultipartEncoderMonitor

NEXUS_URL = os.environ.get("NEXUS_URL", "http://localhost:8081")
USERNAME = os.environ.get("NEXUS_USER", "admin")
PASSWORD = os.environ.get("NEXUS_PASS", "admin")

def collect_files(src: str) -> List[str]:
    file_paths = []
    for root, _, files in os.walk(src):
        for file in files:
            file_paths.append(os.path.join(root, file))
    return file_paths

def upload_files(src: str, repository: str, subdir: Optional[str] = None, quiet: bool = False):
    file_paths = collect_files(src)
    fields = {}
    for idx, file_path in enumerate(file_paths, 1):
        rel_path = os.path.relpath(file_path, src)
        rel_path = rel_path.replace(os.sep, "/")
        fields[f'raw.asset{idx}'] = (os.path.basename(file_path), open(file_path, 'rb'))
        fields[f'raw.asset{idx}.filename'] = (None, rel_path)
    fields['raw.directory'] = (None, subdir)

    upload_endpoint = f"{NEXUS_URL}/service/rest/v1/components?repository={repository}"
    encoder = MultipartEncoder(fields=fields)
    
    # Show progress bar only if stdout is a TTY and not in quiet mode
    show_progress = sys.stdout.isatty() and not quiet
    progress = tqdm(total=encoder.len, unit='B', unit_scale=True, desc='Uploading', disable=not show_progress)

    def callback(monitor):
        progress.update(monitor.bytes_read - progress.n)

    monitor = MultipartEncoderMonitor(encoder, callback)
    headers = {'Content-Type': monitor.content_type}
    response = requests.post(
        upload_endpoint,
        auth=(USERNAME, PASSWORD),
        data=monitor,
        headers=headers
    )
    progress.close()
    for key, value in fields.items():
        if isinstance(value, tuple) and hasattr(value[1], 'close'):
            value[1].close()
    if not quiet:
        if response.status_code == 204:
            print(f"Uploaded {len(file_paths)} files from {src}")
        else:
            print(f"Failed to upload files: {response.status_code} {response.text}")

def main(args):
    if "/" in args.dest:
        repository, subdir = args.dest.split("/", 1)
    else:
        repository, subdir = args.dest, None
    quiet = getattr(args, 'quiet', False)
    upload_files(args.src, repository, subdir, quiet)
